File: OOP/polymorphism.py
class vehicle:
    def __init__(self) -> None:
        pass


class aeroplane(vehicle):
    distance = 0

    def __init__(self, destination, dis) -> None:
        self.going_to = destination
        self.distance = dis
        super().__init__()

    def fair(self):
        print(f"Your flight costs around: {self.distance * 100}")


class bus(vehicle):
    distance = 0

    def __init__(self, destination, dis) -> None:
        self.going_to = destination
        self.distance = dis
        super().__init__()

    def fair(self):
        print(f"Your bus fare costs around: {self.distance * 30}")


class taxi(vehicle):
    distance = 0

    def __init__(self, destination, dis) -> None:
        self.going_to = destination
        self.distance = dis
        super().__init__()

    def fair(self):
        print(f"Your taxi fare costs around: {self.distance * 10}")


person_rich = aeroplane("Cox's Bazar", 150)
person_middle_class = bus("Cox's Bazar", 150)
person_poor_me = taxi("Cox's Bazar", 150)


person_rich.fair()
person_middle_class.fair()
person_poor_me.fair()

File: OOP/polymorphism.cpp
#include <bits/stdc++.h>

using namespace std;

class Vehicle
{
public:
    virtual void fare() const = 0;
};

class Aeroplane : public Vehicle
{
private:
    string going_to;
    int distance;

public:
    Aeroplane(string destination, int dis) : going_to(destination), distance(dis) {}

    void fare() const override
    {
        cout << "Your flight costs around: " << distance * 100 << endl;
    }
};

class Bus : public Vehicle
{
private:
    string going_to;
    int distance;

public:
    Bus(string destination, int dis) : going_to(destination), distance(dis) {}

    void fare() const override
    {
        cout << "Your bus fare costs around: " << distance * 30 << endl;
    }
};

class Taxi : public Vehicle
{
private:
    string going_to;
    int distance;

public:
    Taxi(string destination, int dis) : going_to(destination), distance(dis) {}

    void fare() const override
    {
        cout << "Your taxi fare costs around: " << distance * 10 << endl;
    }
};

int main()
{
    Aeroplane person_rich("Cox's Bazar", 150);
    Bus person_middle_class("Cox's Bazar", 150);
    Taxi person_poor_me("Cox's Bazar", 150);

    person_rich.fare();
    person_middle_class.fare();
    person_poor_me.fare();

    return 0;
}
